sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -30,54 +30,61 @@ import org.apache.spark.sql.types._
  */
 object NestedColumnAliasing {
 
-  def unapply(plan: LogicalPlan): Option[LogicalPlan] = plan match {
+  def unapply(plan: LogicalPlan): Option[Map[Attribute, Seq[ExtractValue]]] = plan match {
```

Comment:
It's a bit unnatural to return an option of map for getAttributeToExtractValues. How about using the unapply method to match the pattern and extract the information needed to build the map: 
```scala
def unapply(plan: LogicalPlan): Option[(Seq[Expression], Seq[Attribute])] = plan match {
  case Project(projectList, g: Generate) if (SQLConf.get.nestedPruningOnExpressions ||
      SQLConf.get.nestedSchemaPruningEnabled) && canPruneGenerator(g.generator) =>
    Some((projectList ++ g.generator.children, g.qualifiedGeneratorOutput))
  case ...
}
```
Then build the mapping in the optimizer rule.
```scala
case p @ NestedColumnAliasing((exprs, excludedAttrs)) =>
  ...
```

Reply:
Right now, `unapply` returns a `None` instead of `Some(Map)` if all the nested fields are used. We won't fall through in this case if we only return the information needed to build a map. Let me change this around to see what would achieve the same behavior.